#163153952 Feature/163153952/login user

habibaudu opened the pull request:
#### What does this PR do?
Enable User to Login

#### Description of Task to be completed?
- Write controller to login user
- write unit Test for failures
- write Unit tests for success
- Ensure User can log in using either Email or Username
- Ensure user can log in with either uppercase or lower case email, username, password
- write validation for invalid credentials

#### How should this be manually tested?
- Fire Up postman 
- URL: http://localhost:3000/api/v1/users/login
- Enter username or email and password.
- click send

#### What are the relevant pivotal tracker stories?
- [#163153952](https://www.pivotaltracker.com/story/show/163153952)

habibaudu committed:
feat(login):get the login route workin
write tests for success
write test for failure
ensure login with either username or email
ensure login with either Uppercase or lowwercse
ensure errors are handled by error handler
[#163153952]

habibaudu committed:
commit changes from develop
[#163153952]

oreoluwade commented on controllers/user.js:
```diff
@@ -66,31 +66,43 @@ class Users {
-        return next(err);
-      }
+  static async login(req, res) {
+    const usernameOrEmail = (req.body.email)
```
The logic here is not quite accurate. Yes, you should get two properties from `req.body` which may be (as you have tried to) ``usernameOrEmail` and then `password`. 

At the point of finding the user with `User.findOne`, what you'd be doing is finding the user when `username` is equal to `req.body.usernameOrEmail` or email is equal to `req.body.usernameOrEmail`.

You'd have to use sequelize operators to accomplish this.

@rajeman already used one sequelize operator in the user validator, so he might be able to provide some context. You'd use two in your case; the `or` operator and the `iLike` operator.

Check sequelize documentation for the use of operators. Also, pair with @rajeman to implement this.

habibaudu replied:
ok... will do

oreoluwade commented on controllers/user.js:
```diff
@@ -66,31 +66,43 @@ class Users {
+    const loginUser = await User
+      .findOne({ where: usernameOrEmail });
+    if (loginUser) {
+      if (bcrypt.compareSync(req.body.password.toLowerCase(),
```
The password should not be mutated with any string method; `toLowerCase`, `toUpperCase` or to anything at all. It should remain as it was supplied, else there won't be a match when comparing passwords.

habibaudu replied:
ok thanks, will change it.

oreoluwade commented on controllers/user.js:
```diff
@@ -66,31 +66,43 @@ class Users {
+    let { username, email, password } = req.body;
+    username = username.toLowerCase();
+    email = email.toLowerCase();
+    password = password.toLowerCase();
```
Same thing here. Don't mutate the password at the point of registration.

oreoluwade commented on middlewares/userValidations.js:
```diff
@@ -109,6 +109,46 @@ export default {
     next();
+  },
+  loginParamsValidator(req, res, next) {
+    const availableParams = (req.body.email)
```
The params you should be checking for while logging in will be `usernameOrEmail` and `password` to rhyme with what is in the controller.

habibaudu replied:
okay... thanks

habibaudu committed:
use sequelize operator in login User
remove .toLowerCase from password and emaill
[#163153952]

oaboluwarin commented on controllers/user.js:
```diff
@@ -66,22 +69,28 @@ class Users {
+      // eslint-disable-next-line max-len
+      .findOne({ where: { [or]: [{ username: { [iLike]: req.body.username } }, { email: { [iLike]: req.body.email } }] } });
+    if (loginUser) {
+      if (bcrypt.compareSync(req.body.password, loginUser.password)) {
```
There is a `validPassword` method available on the Prototype of the `User` model. The function encapsulates the password checking functionality. That's what you should use to check the password here. No need to import `bcrypt`.

TLDR: Check out the `validPassword` method in the `User` model for more context.

habibaudu replied:
ok.thanks.

habibaudu committed:
use password compare from model
[#163153952]

habibaudu committed:
write more test
restructure user login controller
restructure login validation middleware
add isvalidCredential middleware
[#163153952]

oreoluwade commented on controllers/user.js:
```diff
@@ -66,22 +69,22 @@ class Users {
+      });
+    await loginUser.validPassword(password);
+    return res.status(200).json({
+      message: 'Login was sucessful'
```
Please correct the spelling of successful here.

oreoluwade commented on controllers/user.js:
```diff
@@ -66,22 +69,22 @@ class Users {
+    const loginUser = await User
+      .findOne({
+        where: {
+          [or]: [{ username: { [iLike]: usernameOrEmail } },
```
Step the `username` object to the next line with its curly braces so that it aligns with the email object.

oreoluwade commented on controllers/user.js:
```diff
@@ -66,22 +69,22 @@ class Users {
     });
   }
 
+
```
Remove this extra line. One line is enough between two methods.

oreoluwade commented on middlewares/userValidations.js:
```diff
@@ -109,6 +109,69 @@ export default {
+    error.status = 400;
+    return next(error);
+  },
+  async invalidCredential(req, res, next) {
```
Please pluralize this `invalidCredential` middleware

oreoluwade commented on middlewares/userValidations.js:
```diff
@@ -109,6 +109,69 @@ export default {
+    const loginUser = await User
+      .findOne({
+        where: {
+          [or]: [{ username: { [iLike]: usernameOrEmail } },
```
Step the username object down like before.

oreoluwade commented on middlewares/userValidations.js:
```diff
@@ -109,6 +109,69 @@ export default {
+      });
+
+    if (!loginUser) {
+      const error = new Error('Invalid Credential');
```
Pluralize `credential`

oreoluwade commented on middlewares/userValidations.js:
```diff
@@ -109,6 +109,69 @@ export default {
+      error.status = 400;
+      return next(error);
+    }
+    const valid = await loginUser.validPassword(password);
```
Since the validPassword method on the user object returns a boolean, it may be best to make the variable name more inclined towards that. For example, `passwordIsValid`

habibaudu committed:
add a new line after every function and test suite
[#163153952]

oreoluwade reviewed (approved):
LGTM
